ocs_ci/deployment/flexy.py

```diff
@@ -0,0 +1,342 @@
+logger = logging.getLogger(__name__)
+
+
+def flexy_post_processing(func):
+    """
+    Perform copying of flexy-dir to nfs mount
+    and do this only if its jenkins run
+
+    Args:
+        func : run_container function
+    """
+    # Check whether its a jenkins run
+    if is_jenkins_mount():
+        def wrapper(*args, **kwargs):
+            flexy_nfs_path = os.path.join(
+                constants.JENKINS_NFS_CURRENT_CLUSTER_DIR,
+                constants.FLEXY_HOST_DIR
+            )
+            func(*args, **kwargs)
+            if not os.path.exists(flexy_nfs_path):
+                shutil.copytree(
+                    constants.FLEXY_HOST_DIR_PATH,
+                    flexy_nfs_path
+                )
+        return wrapper
+    else:
+        return func
+
+
+def is_jenkins_mount():
+    """
+    Find if this is jenkins run based on current-cluster-dir and
+    NFS mount
+
+    Returns:
+        bool: True if this is jenkins run else False
+
+    """
+    return (
+        os.path.exists(constants.JENKINS_NFS_CURRENT_CLUSTER_DIR)
+        and os.path.ismount(constants.JENKINS_NFS_CURRENT_CLUSTER_DIR)
+    )
+
+
+class FlexyBase(object):
+    """
+    A base class for all types of flexy installs
+    """
+    def __init__(self):
+        self.cluster_name = config.ENV_DATA['cluster_name']
+        # Host dir path which will be mounted inside flexy container
+        # This will be root for all flexy housekeeping
+        self.flexy_host_dir = os.path.expanduser(
+            constants.FLEXY_HOST_DIR_PATH
+        )
+
+        # Path inside container where flexy_host_dir will be mounted
+        self.flexy_mnt_container_dir = config.ENV_DATA.get(
+            'flexy_mnt_container_dir', constants.FLEXY_MNT_CONTAINER_DIR
+        )
+        self.flexy_img_url = config.ENV_DATA.get('flexy_img_url')
+
+        # If user has provided an absolute path for env file
+        # then we don't clone private-conf repo else
+        # we will look for 'flexy_private_conf_url' , if not specified
+        # we will go ahead with default 'flexy-ocs-private' repo
+        if not config.ENV_DATA.get('flexy-env-file'):
+            self.flexy_private_conf_url = config.ENV_DATA.get(
+                'flexy_private_conf_url',
+                constants.FLEXY_DEFAULT_PRIVATE_CONF_URL
+            )
+            self.flexy_private_conf_branch = config.ENV_DATA.get(
+                'flexy_private_conf_branch',
+                constants.FLEXY_DEFAULT_PRIVATE_CONF_BRANCH
+            )
+            # Host dir path for private_conf dir where private-conf will be
+            # cloned
+            self.flexy_host_private_conf_dir_path = os.path.join(
+                self.flexy_host_dir, 'flexy-ocs-private'
+            )
+            self.flexy_env_file = os.path.join(
+                self.flexy_host_private_conf_dir_path, constants.FLEXY_DEFAULT_ENV_FILE
+            )
+        else:
+            self.flexy_env_file = config.ENV_DATA['flexy-env-file']
```

Comment:
I'm thinking about the logic here and the comment and I think that we should slightly change the logic somehow - in this case it will not be possible to have multiple env files in the flexy private repository - because when you will provide different env file name, it will skip the cloning of the private repository.

I would suggest following approach:
* set `self.flexy_env_file` to `ENV_DATA['flexy_env_file']`, if configured or use default from  `constants.FLEXY_DEFAULT_ENV_FILE`
* check if `self.flexy_env_file` exists
* if `self.flexy_env_file` does not exists, expects that it is part of the private flexy repo and clone that repo
* finally once more check, that the `self.flexy_env_file` exists - if not raise exception

Reply:
How about this:
We will check ENV_DATA['flexy_env_file'] whether it's absolute path or relative path (whether it starts with '/' OR not), if its absolute path then we will not do anything (assuming that its run on local machine), if its relative path then we assume its inside private repo. So we can have multiple env files in repository (like independent-conf.env, baremetal-conf.env etc) and just that name will be passed to flexy_env_file. WDYT ?

Reply:
@shylesh I agree with your suggestion, if we will need some enhancement, we can do it in separated PR later.